ServerCore/Pages/Puzzles/Edit.cshtml.cs

```diff
@@ -120,6 +120,41 @@ public async Task<IActionResult> OnPostAsync(int puzzleId)
             {
                 await _context.SaveChangesAsync();
 
+                // Check to see if IsForSinglePlayer has changed.
+                // Delete appropriate rows if so
+                if (oldPuzzleView.IsForSinglePlayer != Puzzle.IsForSinglePlayer)
+                {
+                    if (oldPuzzleView.IsForSinglePlayer)
+                    {
+                        // If moving from SinglePlayerPuzzle to team puzzle, delete single player puzzle states
+                        IEnumerable<SinglePlayerPuzzleHintStatePerPlayer> hintStatesToRemove = _context.SinglePlayerPuzzleHintStatePerPlayer.Where(hintState => hintState.Hint.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleHintStatePerPlayer.RemoveRange(hintStatesToRemove);
+
+                        IEnumerable<SinglePlayerPuzzleStatePerPlayer> puzzleStatesToRemove = _context.SinglePlayerPuzzleStatePerPlayer.Where(puzzleState => puzzleState.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleStatePerPlayer.RemoveRange(puzzleStatesToRemove);
+
+                        IEnumerable<SinglePlayerPuzzleSubmission> submissionsToRemove = _context.SinglePlayerPuzzleSubmissions.Where(submissions => submissions.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleSubmissions.RemoveRange(submissionsToRemove);
+
+                        IEnumerable<SinglePlayerPuzzleUnlockState> unlockStatesToRemove = _context.SinglePlayerPuzzleUnlockStates.Where(unlockState => unlockState.PuzzleID == oldPuzzleView.ID);
+                        _context.SinglePlayerPuzzleUnlockStates.RemoveRange(unlockStatesToRemove);
+                    }
+                    else
+                    {
+                        // If moving from team puzzle to SinglePlayerPuzzle, delete team puzzle states
+                        IEnumerable<HintStatePerTeam> hintStatesToRemove = _context.HintStatePerTeam.Where(hintState => hintState.Hint.PuzzleID == oldPuzzleView.ID);
+                        _context.HintStatePerTeam.RemoveRange(hintStatesToRemove);
+
+                        IEnumerable<PuzzleStatePerTeam> puzzleStatesToRemove = _context.PuzzleStatePerTeam.Where(puzzleState => puzzleState.PuzzleID == oldPuzzleView.ID);
+                        _context.PuzzleStatePerTeam.RemoveRange(puzzleStatesToRemove);
+
+                        IEnumerable<Submission> submissionsToRemove = _context.Submissions.Where(submissions => submissions.PuzzleID == oldPuzzleView.ID);
+                        _context.Submissions.RemoveRange(submissionsToRemove);
+                    }
+
```

Comment:
We also need to create the appropriate rows in the database for the new puzzle type.  If it's now a single player puzzle, SinglePlayerPuzzleUnlockStates needs to be created, and if it's now a team puzzle, PuzzleStatePerTeam needs to be created for every team.  Mirroring the logic in https://github.com/PuzzleServer/mainpuzzleserver/blob/main/ServerCore/Pages/Puzzles/Create.cshtml.cs#L90-L103 should be sufficient.

Reply:
+1

Reply:
Or is this what new lines 218-228 are doing?

Reply:
It is what those lines were doing